src/adapter/src/coord/peek.rs

```diff
@@ -423,6 +493,102 @@ impl crate::coord::Coordinator {
 
         let timestamp = determination.timestamp_context.timestamp_or_default();
 
+        if let PeekPlan::FastPath(FastPathPlan::PeekPersist(id, mfp_plan)) = fast_path {
+            let mut cursor = self
+                .controller
+                .storage
+                .snapshot_cursor(id, timestamp)
+                .await?;
+
+            let metrics = self.metrics.clone();
+            let handle: JoinHandle<Result<PeekResponseUnary, String>> =
+                mz_ore::task::spawn(|| "persist::peek", async move {
+                    let mut limit_remaining = finishing.limit.unwrap_or(usize::MAX);
+
+                    // Re-used state for processing and building rows.
+                    let mut accum = vec![];
+                    let mut datum_vec = DatumVec::new();
+                    let mut row_builder = Row::default();
+                    let arena = RowArena::new();
+
+                    let start = Instant::now();
+                    let mut last_fetch = Duration::ZERO;
+                    while limit_remaining > 0 {
+                        let Some(batch) = cursor.next().await else {
+                            break;
+                        };
+                        last_fetch = start.elapsed();
+                        for ((k, v), _, d) in batch {
+                            let row = k?.0.map_err(|e| e.to_string())?;
+                            let () = v?;
+                            let count: usize = d
+                                .try_into()
+                                .map_err(|_| "Negative count for thing in snapshot".to_owned())?;
+                            let Some(count) = NonZeroUsize::new(count) else {
+                                continue;
+                            };
+                            let mut datum_local = datum_vec.borrow_with(&row);
+                            let eval_result = mfp_plan
+                                .evaluate_into(&mut datum_local, &arena, &mut row_builder)
+                                .map_err(|e| e.to_string())?;
+                            if let Some(row) = eval_result {
+                                accum.push((row, count));
+                                limit_remaining = limit_remaining.saturating_sub(count.get());
+                                if limit_remaining == 0 {
+                                    break;
+                                }
+                            }
+                        }
+                    }
```

Comment:
Yes, running in environmentd!

Is it okay if I take this as a followup? I would like to roll this change in with some of the other limit-enforcement stuff mentioned in the PR description to make it easier to test. In the meantime this will stay guarded behind the flag.

Reply:
Sure, would you mind adding a comment or issue so it doesn't slip through the cracks?

Reply:
Certainly: added to the task list at https://github.com/MaterializeInc/materialize/issues/21265